enhancements/operator-dependency-resolution.md

```diff
@@ -0,0 +1,222 @@
+* OpenShift version (OpenShift only)
+
+After evaluating several libraries, we have chosen https://github.com/irifrance/gini as the SAT solver. This implementation of the solver is general, a popular library, and is written in pure go.
+
+#### Distributed Problem
+
+We also need to consider the performance impact of how the sat solver's dependencies are aggregated. The challenge with distributed catalogs is that the data isn't locally available and the path to resolution will by default create significant load on the existing registry pods for each query. We should attempt to limit this wherever possible and, based on profiling after doing some initial work building the resolver, build a resolution cache of all existing catalog at runtime. The initial design for that cache would require aggregating all data in the registry that the resolver cares about and preparing it for a single query, then having the catalog operator fetch it once whenever the registry is updated.
+
+#### Library
+
+The resolver will be a library that can be imported not just by OLM and used at runtime. This will be a separate component that can be used to determine how OLM will resolve a dependency outside the scope of a running cluster if need be. As a result, we will create a separate git repository for the resolver and have OLM import it.
+
+### Dependency Specification
+
+The dependencies of an operator are listed as a list in `dependencies.yaml` file inside `/metadata` folder of a bundle. This file is optional and only used to specify explicit operator version dependencies at first. Eventually, operator authors can migrate the API-based dependencies into `dependencies.yaml` as well in the future. The ultimate goal is to have `dependencies.yaml` as a centralized metadata for operator dependencies and moving the dependency information away from CSV.
+
+The registry will parse the dependency information in `dependencies.yaml` and add to the database when the bundle is loaded into the registry. OLM will query the dependency information from the registry to facilitate the dependency resolution using sat solver.
+
+### Syntax
+
+The dependency list will contain a `type` field for each item to specify what kind of dependency this is. It can be a package type (`olm.package`) meaning this is a dependency for a specific operator. For `olm.package` type, the dependency information should include the `package` name and the `version` of the package in semver format. We use `blang/semver` library for semver parsing (https://github.com/blang/semver). For example, you can specify an exact version such as `0.5.2` or a range of version such as `>0.5.1` (https://github.com/blang/semver#ranges). In addition, the author can specify dependency that is similiar to existing CRD/API-based using proposed `olm.gvk` type and then specify GVK information as how it is done in CSV. This is a path to enable operator authors to consolidate all dependencies (API or explicit) to be in the same place.
```

Comment:
After we do this initial pass, I think we'll do one of two things with channels:

1. Channels stay as they are today, and we implement your suggestion to allow specifying channels as part of your dependencies.
2. Things that people are using channels for today are replaced by other constraints specified by the author and given to the resolver (openshift version, supported operand version, single/multi namespace etc), and we remove channels entirely or we make them strictly for update speed (i.e. "stable" and "fast")

Reply:
IMO, having channel level control would complete this spec.  If OLM doesn't assume the default channel, I think it would improve the author's experience at no cost to the user because authors would have one less thing to worry about since the user can look through channels for their desired version.  In addition, I don't think assuming the default channel will work with y-stream, x-stream, version-stream type channels as dependencies.  If myApp depends on istio version <2.1, istio's default channel is 2.0 (y-stream channel), and an upgrade comes in for istio with the default channel now 2.1, I'll miss out on new z-stream versions. 

I'm curious about the timeline on the channel discussion.  Is it something that can be done in parallel to this spec? If not, I think it would it be worth adding channel level control until this is sorted.

Reply:
Hey Ryan,

I'm curious, why would defining the channel explicitly be necessary for the dependency? Is this because of the case where you have different package versions that share the same "version" currently?

To me, the purpose of a cluster admin selecting the channel should be equivalent to them saying "this is how frequently we want updates". That is definitely outside the scope of what an operator developer should care about. They should just care that "my thing only works on this set of versions" however that happens to be defined

Reply:
Hey Kevin,

> I'm curious, why would defining the channel explicitly be necessary for the dependency? 
>
With version-streamed dependencies, the versions are built into the channel name.  So when additional channels come around with newer versions, the defaultChannel name changes, changing the pool of available versions for OLM to pick from.  That can cause irregular behavior during upgrade.

> Is this because of the case where you have different package versions that share the same "version" currently?
>
I'm not sure where to check package version.  Is that the name of the CSV used in the package?  If that's the case,  no, the version is included in the name's string so it's always different.

> the purpose of a cluster admin selecting the channel should be equivalent to them saying "this is how frequently we want updates"
>
I guess I have a different expectation when communication channel types.  Something like stable to me would be "I'm getting ga releases".  While something like 2.0 would be "I'm getting all z-streams for version 2.0".

Reply:
> With version-streamed dependencies, the versions are built into the channel name. So when additional channels come around with newer versions, the defaultChannel name changes, changing the pool of available versions for OLM to pick from. That can cause irregular behavior during upgrade.

So I guess what I'm saying is that we shouldn't do this. It's kind of abusing the channel concept because of a lack of other available features to do it with. The index images solve this problem for us, because there will just be a separate curated content set (a different index) to tie operator bundle versions to orthogonal software release cycles

> I'm not sure where to check package version. Is that the name of the CSV used in the package? If that's the case, no, the version is included in the name's string so it's always different.

Part of the problem is that this dependency resolution enhancement is making the assumption that a semantic version is tied to a particular operator bundle. See: https://github.com/operator-framework/enhancements/blob/master/enhancements/implicit-catalog-versioning.md If the operator bundle's version (and not the CSV) is used to define a dependency then they can be defined uniquely without the need to define a channel at all. Selecting the default channel for an update stream becomes totally divorced in concept at that point.

> I guess I have a different expectation when communication channel types. Something like stable to me would be "I'm getting ga releases". While something like 2.0 would be "I'm getting all z-streams for version 2.0".

I agree with this. Technically it is possible to define this on your operator bundle right now if you craft the channels the right way, but it's not obvious to all users how to do it. I think making changes to the way channels are defined (as Evan discussed above) is something we want to do in the future to make it more obvious or implicit, but to me it's outside the scope of the dependency resolver specifically.

Reply:
> Part of the problem is that this dependency resolution enhancement is making the assumption that a semantic version is tied to a particular operator bundle
>
I see the disconnect, I haven't made the jump to index images yet, so my assumption was different.  After doing some catch up, I think we're on the same page -- channels aren't what they were previously.